dashboard/modules/reporter/reporter_head.py

```diff
@@ -112,6 +147,233 @@ async def get_cluster_status(self, req):
+
+        except ValueError as e:
+            raise aiohttp.web.HTTPInternalServerError(text=str(e))
+        if not reply.success:
+            return aiohttp.web.HTTPInternalServerError(text=reply.output)
+
+        logger.info("Returning stack trace, size {}".format(len(reply.output)))
+
+        task_ids_in_a_worker = await self.get_task_ids_running_in_a_worker(worker_id)
+        return aiohttp.web.Response(
+            text=WARNING_FOR_MULTI_TASK_IN_A_WORKER
+            + str(task_ids_in_a_worker)
+            + "\n"
+            + reply.output
+            if len(task_ids_in_a_worker) > 1
+            else reply.output
+        )
+
+    @routes.get("/task/cpu_profile")
+    async def get_task_cpu_profile(self, req) -> aiohttp.web.Response:
+        """
+        Retrieves the CPU profile for a specific task.
+        Note that one worker process works on one task at a time or one worker works on multiple async tasks.
+
+        Args:
+            req (aiohttp.web.Request): The HTTP request object.
+
+        Returns:
+            aiohttp.web.Response: The HTTP response containing the CPU profile data.
+
+        Raises:
+            ValueError: If the "task_id" parameter is missing in the request query.
+            ValueError: If the "attempt_number" parameter is missing in the request query.
+            ValueError: If the maximum duration allowed is exceeded.
+            ValueError: If the worker begins working on another task during the profile retrieval.
+            aiohttp.web.HTTPInternalServerError: If there is an internal server error during the profile retrieval.
+            aiohttp.web.HTTPInternalServerError: If the CPU Flame Graph information for the task is not found.
+        """
+        if "task_id" not in req.query:
+            raise ValueError("task_id is required")
+        if "attempt_number" not in req.query:
+            raise ValueError("task's attempt number is required")
+        if "node_id" not in req.query:
+            raise ValueError("node_id is required")
+
+        task_id = req.query.get("task_id")
+        attempt_number = req.query.get("attempt_number")
+        node_id = req.query.get("node_id")
+
+        ip = DataSource.node_id_to_ip[node_id]
+
+        try:
+            (pid, _) = await self.get_worker_details_for_task(task_id, attempt_number)
+        except ValueError as e:
+            raise aiohttp.web.HTTPInternalServerError(text=str(e))
+
+        duration = int(req.query.get("duration", 5))
+        if duration > 60:
+            raise ValueError(f"The max duration allowed is 60: {duration}.")
+        format = req.query.get("format", "flamegraph")
+
+        # Default not using `--native` for profiling
+        native = req.query.get("native", False) == "1"
+        reporter_stub = self._stubs[ip]
+
+        logger.info(
+            "Sending CPU profiling request to {}:{} for {} with native={}".format(
+                ip, pid, task_id, native
+            )
+        )
+
```

Comment:
I wonder if we should also get `get_task_ids_running_in_a_worker` here? 
I know it's hard to know for sure if other tasks are running while we are doing profiling/backtrace, but checking before/after we do profiling/traceback is probably the best efforts.

Reply:
That's a good catch! 

Do you think we move `get_worker_details_for_running_task` just before the  logger and `reporter_stub.CpuProfiling` would be good enough?

Like:


```
try:
          (pid, _) = await self.get_worker_details_for_running_task(
                task_id, attempt_number
            )
        except ValueError as e:
            raise aiohttp.web.HTTPInternalServerError(text=str(e))

        logger.info(
            "Sending CPU profiling request to {}:{} for {} with native={}".format(
                ip, pid, task_id, native
            )
        )

        reply = await reporter_stub.CpuProfiling(
            reporter_pb2.CpuProfilingRequest(
                pid=pid, duration=duration, format=format, native=native
            )
        )

        # Call the API again to make sure task is still running and the worker is still working on the task
        # Since there is a task scheduling strategy that a worker may run different tasks at different time
        try:
            (_, worker_id) = await self.get_worker_details_for_running_task(
                task_id, attempt_number
            )
        except ValueError as e:
            raise aiohttp.web.HTTPInternalServerError(text=str(e))

```

Reply:
it's still best effort - but better I think.